hazelcast/include/hazelcast/client/Pipelining.h

```diff
@@ -0,0 +1,183 @@
+         * of the Pipelining because the backpressure on the client/member is leading.
+         *
+         * The Pipelining has been marked as Beta since we need to see how the API needs to
+         * evolve. But there is no problem using it in production. We use similar techniques
+         * to achieve high performance.
+         *
+         * @param <E>
+         */
+        template<typename E>
+        class Pipelining : public boost::enable_shared_from_this<Pipelining<E> > {
+        public:
+            /**
+             * Creates a Pipelining with the given depth.
+             *
+             * We use this factory create method and hide the constructor from the user, because the private
+             * {@link up()} and {@link down()} methods of this class may be called from the executor thread, and
+             * we need to make sure that the `Pipelining` instance is not destructed when these methods are accessed.
+             *
+             * @param depth the maximum number of concurrent calls allowed in this Pipelining.
+             * @throws IllegalArgumentException if depth smaller than 1. But if you use depth 1, it means that
+             *                                  every call is sync and you will not benefit from pipelining at all.
+             */
+            static boost::shared_ptr<Pipelining> create(int depth) {
+                util::Preconditions::checkPositive(depth, "depth must be positive");
+
+                return boost::shared_ptr<Pipelining>(new Pipelining(depth));
+            }
+
+            /**
+             * Returns the results.
+             * <p>
+             * The results are returned in the order the requests were done.
+             * <p>
+             * This call waits till all requests have completed.
+             *
+             * @return the List of results.
+             * @throws IException if something fails getting the results.
+             */
+            std::vector<boost::shared_ptr<E> > results() {
+                std::vector<boost::shared_ptr<E> > result(futures.size());
+                size_t index = 0;
+                for (typename FuturesVector::const_iterator it = futures.begin(); it != futures.end(); ++it, ++index) {
+                    result[index] = (*it)->get();
+                }
+                return result;
+            }
+
+            /**
+             * Adds a future to this Pipelining or blocks until there is capacity to add the future to the Pipelining.
+             * <p>
+             * This call blocks until there is space in the Pipelining, but it doesn't mean that the invocation that
+             * returned the ICompletableFuture got blocked.
+             *
+             * @param future the future to add.
+             * @return the future added.
+             * @throws NullPointerException if future is null.
+             */
+            const boost::shared_ptr<ICompletableFuture<E> >
```

Comment:
We should either return non-const shared ptr, or const shrd_ptr & . 
Returning const and copy does not cause any problem but it is confusing.

Reply:
OK, changed to `const boost::shared_ptr<ICompletableFuture<E> > &`